src/api/trace/span_context.rs

```diff
@@ -111,6 +114,110 @@ impl SpanId {
     }
 }
 
+/// TraceState carries system-specific configuration data, represented as a list
+/// of key-value pairs. TraceState allows multiple tracing systems to
+/// participate in the same trace.
+///
+/// Please review the [W3C specification] for details on this field.
+///
+/// [W3C specification]: https://www.w3.org/TR/trace-context/#tracestate-header
+#[cfg_attr(feature = "serialize", derive(Deserialize, Serialize))]
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct TraceState {
+    data: HashMap<String, String>,
+    ordered_keys: VecDeque<String>,
+}
+
+impl TraceState {
+    /// Creates a new `TraceState` from the given key-value collection.
+    pub fn from_key_value(trace_state: Vec<KeyValue>) -> Self {
+        let mut data: HashMap<String, String> = HashMap::with_capacity(trace_state.len());
+        let mut ordered_keys: VecDeque<String> = VecDeque::with_capacity(trace_state.len());
+
+        for kv in trace_state {
+            data.insert(kv.key.clone().into(), kv.value.into());
+            ordered_keys.push_back(kv.key.into());
+        }
+
+        TraceState { data, ordered_keys }
+    }
+
+    /// Retrieves a value for a given key from the `TraceState` if it exists.
+    pub fn get(&self, key: String) -> Option<String> {
+        self.data.get(&key).cloned()
+    }
+
+    /// Inserts the given key-value pair into the `TraceState`. If a value already exists for the
+    /// given key, this updates the value and updates the value's position. Returns the updated
+    /// `TraceState`.
+    pub fn insert(mut self, key: String, value: String) -> TraceState {
+        if let Some(index) = self.ordered_keys.iter().position(|x| *x == *key) {
+            self.ordered_keys.remove(index);
+        }
+
+        self.data.insert(key.clone(), value);
+        self.ordered_keys.push_front(key);
+
+        self
+    }
+
+    /// Removes the given key-value pair from the `TraceState`, returning the removed value if it
+    /// exists.
+    pub fn delete(mut self, key: String) -> TraceState {
+        if let Some(index) = self.ordered_keys.iter().position(|x| *x == *key) {
+            self.ordered_keys.remove(index);
+            self.data.remove(&key);
+        }
+
+        self
+    }
+
+    /// Creates a new `TraceState` header string, delimiting each key and value with a `=` and each
+    /// entry with a `,`.
+    pub fn header(&self) -> String {
+        self.header_delimited("=", ",")
+    }
+
+    /// Creates a new `TraceState` header string, with the given key/value delimiter and entry delimiter.
+    pub fn header_delimited(&self, entry_delimiter: &str, list_delimiter: &str) -> String {
+        let mut ordered: Vec<String> = Vec::with_capacity(self.ordered_keys.len());
+        for key in self.ordered_keys.clone() {
+            ordered.push(format!(
+                "{}{}{}",
+                key,
+                entry_delimiter,
+                self.data.get(key.as_str()).cloned().unwrap()
```

Comment:
could do `as_ref().unwrap()` here to avoid the clone.

If you go with `BTreeMap` this also could be simplified to:

```rust
for (key, value) in self.data.iter() {
    ordered.push(format!("{}{}{}", key, entry_delimiter, value))
}
```